analyzers/src/SonarAnalyzer.CSharp/Rules/ParameterTypeShouldMatchRouteTypeConstraint.cs

```diff
@@ -0,0 +1,238 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2023 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using Microsoft.CodeAnalysis.Text;
+
+namespace SonarAnalyzer.Rules.CSharp;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public sealed class ParameterTypeShouldMatchRouteTypeConstraint : SonarDiagnosticAnalyzer
+{
+    private const string DiagnosticId = "S6800";
+    private const string MessageFormat = "{0}";
+    private const string ImplicitStringPrimaryMessageFormat = "Parameter type '{0}' does not match route parameter implicit type constraint 'string' in route '{1}'.";
+    private const string PrimaryMessageFormat = "Parameter type '{0}' does not match route parameter type constraint '{1}' in route '{2}'.";
+    private const string MessageWithSecondaryLocationFormat = "Parameter type '{0}' does not match route parameter type constraint.";
+    private const string SecondaryMessageFormat = "This route parameter has a '{0}' type constraint.";
+    private const string ImplicitStringSecondaryMessage = "This route parameter has an implicit 'string' type constraint.";
+
+    private const string ImplicitStringConstraint = "string";
+
+    private static readonly DiagnosticDescriptor Rule = DescriptorFactory.Create(DiagnosticId, MessageFormat);
+
+    // https://learn.microsoft.com/en-us/aspnet/core/blazor/fundamentals/routing?view=aspnetcore-7.0#route-constraints
+    private static readonly Dictionary<string, SupportedType> ConstraintMapping = new(StringComparer.InvariantCultureIgnoreCase)
```

Comment:
Possible optimization:
* Move `ConstraintMapping` to compilation start and capture it.
* Replace the value `new SupportedType(..)` by `Func<ITypeSymbol, bool>` (Implemented like `Matches`)

In the CompliationStartAction:
```cs
var constraintMapping = new Dictionary<string, Func<ITypeSymbol, bool>>(StringComparer.InvariantCultureIgnoreCase)
{
    { "bool", MatchFunction(KnownType.System_Boolean),
};

Func<ITypeSymbol, bool> MatchFunction(KnownType knownType)
    => MatchFunction(cc.Compilation, cc.Compilation.GetTypeByMetadataName(knownType));

static Func<ITypeSymbol, bool> MatchFunction(Compilation compilation, ITypeSymbol targetType)
    => (ITypeSymbol expressionType) =>
{
// see the old implementation
}
```

That way, 
* you do not pass the compilation to `match`
* you only call `GetTypeByMetadataName` once at the compilation start

But you have to pass the dictionary around. I would consider it, but leave it as is if it is too cumbersome.

Reply:
I would propose to have this in a dedicated PR, and merge this PR without this optimization, so that we can also unblock https://github.com/SonarSource/sonar-dotnet/issues/8275.